push.sh

```diff
@@ -1,16 +1,72 @@
 #! /bin/sh
 
-# Use WEBDAV_USERNAME or WEBDAV_PASSWORD as default, if provided
+# Use WEBDAV_USERNAME as default, if provided.
 if [ -z "$PLUGIN_USERNAME" ] && [ ! -z "$WEBDAV_USERNAME" ]; then
+
     PLUGIN_USERNAME="$WEBDAV_USERNAME"
 fi
+
+# Use WEBDAV_PASSWORD as default, if provided.
 if [ -z "$PLUGIN_PASSWORD" ] && [ ! -z "$WEBDAV_PASSWORD" ]; then
+
     PLUGIN_PASSWORD="$WEBDAV_PASSWORD"
 fi
 
 # If username and password are provided, add auth
 if [ ! -z "$PLUGIN_USERNAME" ] && [ ! -z "$PLUGIN_PASSWORD" ]; then
-    AUTH="-u ${PLUGIN_USERNAME}:${PLUGIN_PASSWORD}"
+
+    AUTH="--user ${PLUGIN_USERNAME}:${PLUGIN_PASSWORD}"
+fi
+
+# Use a proxy, if one is specified
+if [ ! -z "$PLUGIN_PROXY_URL" ]; then
+
+    PLUGIN_PROXY_URL="--proxy ${PLUGIN_PROXY_URL}"
 fi
 
-curl -T $PLUGIN_FILE $AUTH $PLUGIN_DESTINATION
+# If a timeout is specified, make use of it. If none is defined, use a default value (30 seconds).
+if [ ! -z "$PLUGIN_TIMEOUT" ]; then
+
+    PLUGIN_TIMEOUT="--max-time ${PLUGIN_TIMEOUT}"
+else
+
+    PLUGIN_TIMEOUT="--max-time 1800"
+fi
+
+# Set PLUGIN_ATTEMPTS to one if nothing else is specified
+if [ -z "$PLUGIN_ATTEMPTS" ]; then
+
+    PLUGIN_ATTEMPTS=1
+fi
+
+# Repeat the upload as long as specified.
+while [ $PLUGIN_ATTEMPTS -gt 0 ]; do
+
+  # Uploading the file
+  curl $PLUGIN_PROXY_URL --upload-file $PLUGIN_FILE $AUTH $PLUGIN_DESTINATION --progress-bar $PLUGIN_TIMEOUT
+
+  # Terminate the script as soon as the upload is successful
+  if [ $? -eq 0 ]; then
+
+    echo "[INFO] Upload was successful."
+    exit 0
+
+  else
+
+	# Show messages in case uploads have failed
+	if [ $PLUGIN_ATTEMPTS -gt 1 ]; then
+
+		echo "[INFO] Upload failed. Attempting a new upload, if possible."
+	else
+
+		echo "[INFO] All upload attempts have failed."
```

Comment:
[minor] This seems like it should be `[ERROR]` as it is a failure result. We can also `exit 1` here.

Reply:
Does this not happen automatically in line 72?

Reply:
It does, yes. However if refactoring later it could become a possible bug if missed. If at this point we know we have failed and do not wish to continue, why continue evaluating the script?

Reply:
Here's a simpler idea... We can move the `else` condition to after the while loop so that it gets logged just before the `exit 1`. That way all failure logic is contained rather than split in two places.